src/model/mod.rs

```diff
@@ -48,8 +50,145 @@ pub enum PlayingItem {
     Episode(show::FullEpisode),
 }
 
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct Id<'id> {
+    _type: Type,
+    id: &'id str,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy, Display, Error)]
+pub enum IdError {
+    InvalidPrefix,
+    InvalidFormat,
+    InvalidType,
+    InvalidId,
+}
+
+impl std::fmt::Display for Id<'_> {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        f.write_str("spotify:")?;
+        f.write_str(self._type.as_ref())?;
+        f.write_str(":")?;
+        f.write_str(self.id)?;
+        Ok(())
+    }
+}
+
+impl AsRef<str> for Id<'_> {
+    fn as_ref(&self) -> &str {
+        self.id
+    }
+}
+
+impl Id<'_> {
+    pub fn _type(&self) -> Type {
+        self._type
+    }
+
+    pub fn id(&self) -> &str {
+        &self.id
+    }
+
+    pub fn uri(&self) -> String {
+        format!("spotify:{}:{}", self._type.as_ref(), self.id)
+    }
+
+    pub fn from_id_or_uri<'a, 'b: 'a>(_type: Type, id_or_uri: &'b str) -> Result<Id<'a>, IdError> {
+        match Self::from_uri(id_or_uri) {
+            Ok(id) if id._type == _type => Ok(id),
+            Ok(_) => Err(IdError::InvalidType),
+            Err(IdError::InvalidPrefix) => Self::from_id(_type, id_or_uri),
+            Err(error) => Err(error),
+        }
+    }
+
+    pub fn from_id<'a, 'b: 'a>(_type: Type, id: &'b str) -> Result<Id<'a>, IdError> {
+        if id.chars().all(|ch| ch.is_ascii_alphanumeric()) {
```

Comment:
I think you can use `all(char::is_ascii_alphanumeric)` directly here.

Reply:
No, I can't. `all()` requires `fn(&char) -> bool`, and `char::is_ascii_alphanumeric()` is `fn(char) -> bool`.

Reply:
Ah okay, I didn't realize the `&` in the type signature. Don't worry about it then.